File: Matrix completion/Python/generate_matrix.py
import numpy as np

"""
Code taken from R2RILS by Jonathan Bauch, Boaz Nadler and Pini Zilber
"""

def generate_matrix(n1, n2,  singular_values):
    """
    Generate a matrix with specific singular values
    :param int n1: number of rows
    :param int n2: number of columns
    :param list singular_values: required singular values
    """
    rank = len(singular_values)
    U = np.random.randn(n1, rank)
    V = np.random.randn(n2, rank)
    U = np.linalg.qr(U)[0]
    V = np.linalg.qr(V)[0]
    return U @ np.diag(singular_values) @ V.T

File: Matrix completion/MATLAB/generate_matrix.m
function [X, U, V] = generate_matrix(n1,n2,singular_values)
%
% INPUT:  n1,n2 = number of rows and columns
%         singular_values = non-zero singular values

% OUTPUT: X = n1 x n2 matrix of rank r
%         U = n1 x r left singular vectors
%         V = n2 x r right singular vectors

r = length(singular_values); 
D = diag(singular_values);    % diagonal r x r matrix
Z = randn(n1,r); 

[U, ~, ~] = svd(Z,'econ'); 

Z = randn(n2,r); 
[V, ~, ~] = svd(Z,'econ'); 

X = U * D * V';
